internal/compiler-interface/src/main/java/xsbti/AnalysisCallback2.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright Scala Center, Lightbend, and Mark Harrah
+ *
+ * Licensed under Apache License 2.0
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package xsbti;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Extension to AnalsysisCallback.
+ * This is a compatibility layer created for Scala 3.x compilers.
+ * Even though compiler-interface is NOT intended to be a public API,
+ * Scala 3 compiler publishes their own compiler bridge against Zinc de jour.
+ * By having this interface, they can test if `callback` supports AnalysisCallback2.
+ */
+public interface AnalysisCallback2 extends AnalysisCallback {
+    /**
+     * Register a compilation problem.
+     *
+     * This error may have already been logged or not. Unreported problems may
+     * happen because the reporting option was not enabled via command-line.
+     *
+     * @param what The headline of the error.
+     * @param pos At a given source position.
+     * @param msg The in-depth description of the error.
+     * @param severity The severity of the error reported.
+     * @param reported Flag that confirms whether this error was reported or not.
+     * @param rendered If present, the string shown to the user when displaying this Problem.
+     * @param diagnosticCode The unique code attached to the diagnostic being reported.
+     * @param diagnosticRelatedInformation The possible releated information for the diagnostic being reported.
+     * @param actions Actions (aka quick fixes) that are able to either fix or address the issue that is causing this problem.
+     */
+    void problem2(String what,
```

Comment:
Is this something that the Scala 3 bridge also needs to call? I thought we had all the pieces in place already /cc @ckipp01 
Also couldn't this interface take an xsbti.Problem so we don't need a new one every time there's a new field?

Reply:
Nope, I don't believe this is necessary for Scala 3. These changes were specific to the Scala 2 implementation. You can actually test the actions already with the nightly version of the compiler and the latest metals release and see that everything is working as expected in Scala 3.

Reply:
But the Scala 3 bridge does call `problem`: https://github.com/lampepfl/dotty/blob/ce1ce99b6b2e67d4d2ca05f6732e3eb0ee692f1f/sbt-bridge/src/dotty/tools/xsbt/CompilerBridgeDriver.java#L139
Doesn't that need to become `problem2` too?

Reply:
Huh, I guess it would seem so. I guess I don't fully understand how this is all working or when this `run` is called that you have linked since at least via BSP with sbt current the actions are forwarded as expected.

Reply:
confirmed

Reply:
sbt uses `xsbi.Problem` in 2 different ways:
1. via `xsbi.Reporter` which you get a live callback per compiler warnings
2. via `CompileAnalysis`, `ReadSourceInfos`, `SourceInfo`, which contains list of `xsbti.Problem`

This bug fix pertains the second pathway. I think the initial implementation of LSP or BSP use to resend all errors and warnings for all files at the end of `compile` as a workaround to make sure compiler warnings do not disappear, then in 2022 a few folks realized we're bombarding Metals with unnecessary problems, and we got fixes:
- https://github.com/sbt/sbt/pull/6847 by @tanishiking
- https://github.com/sbt/sbt/pull/6929 by @kpodsiad 

I usually ask @adpi2 to figure out the BSP handling, so I only occasionally jump in so I am not sure if this double dipping is unavoidable, but as a principle, it seems like a good idea that the problems we collect in Analysis contains new problem informations (like CodeAction and DiagnosticRelatedInformation).

Reply:
> it seems like a good idea that the problems we collect in Analysis contains new problem informations (like CodeAction and DiagnosticRelatedInformation).

Agreed. In the BSP reporter, we still use the `CompileAnalysis` to report the warnings from the files that the incremental compiler skipped. It's good to have the code actions in there.